py/farm_ng/core/events_file_reader.py

```diff
@@ -104,52 +119,46 @@ def close(self) -> bool:
         return self.is_closed()
 
     def num_frames(self, uri: Uri) -> int:
+        if not self.offsets:
+            self._compute_offsets()
         if uri.string() not in self.offsets.keys():
             return -1
         return len(self.offsets[uri.string()])
 
     def seek(self, uri: Uri, frame_id: int) -> None:
         assert uri.string() in self.offsets.keys()
         assert frame_id < self.num_frames(uri)
-        self._file_stream.seek(self.offsets[uri.string()][frame_id])
+        maybe_file_stream = self._file_stream
+        if maybe_file_stream is None:
+            return
+        file_stream = cast(IO, maybe_file_stream)
+        file_stream.seek(self.offsets[uri.string()][frame_id])
 
-    def seek_and_read(self, uri: Uri, frame_idx: int) -> Optional[Any]:
-        self.seek(uri, frame_idx)
-        return self.read()
-
-    def read(self) -> Optional[Any]:
+    def read(self) -> Optional[Tuple[event_pb2.Event, Any]]:
         # Read the message's length as bytes and convert it to an integer
-        event_len = int.from_bytes(self._file_stream.read(1), sys.byteorder)
+        maybe_file_stream = self._file_stream
+        if maybe_file_stream is None:
+            return None
+        file_stream = cast(IO, maybe_file_stream)
+
+        event_len = int.from_bytes(file_stream.read(4), sys.byteorder)
         if event_len == 0:
             self.close()
             return None
 
         # Read that number of bytes as the message bytes
-        event: bytes = self._file_stream.read(event_len)
+        event: bytes = file_stream.read(event_len)
 
         event_proto = event_pb2.Event()
         event_proto.ParseFromString(event)
 
         # parse the message
-        message_cls = getattr(
-            importlib.import_module(event_proto.module), event_proto.name
-        )
+        name, package = parse_protobuf_descriptor(event_proto.uri.scheme)
+        message_cls = getattr(importlib.import_module(package), name)
```

Comment:
@ethanrublee not sure if this what you meant at all. I checked the descriptors api but i couldn't find a way to dynamically create descriptors/classes using the native protobuf api.

Reply:
yes this was what I was thinking.